scala/data_structures/bloom_filter.scala

```diff
@@ -0,0 +1,55 @@
+package data_structures
+
+import scala.util.hashing.MurmurHash3
+import scala.util.Random
+
+case class BloomFilter[A](
+  nbOfItems: Int,
+  falsePositiveProbability: Double,
+  private val array: Array[Boolean],
+  private val hashFunctions: List[String => Int]
+) {
+
+  def +=(elem: A): BloomFilter[A] = {
+    val str = elem.toString
+    val hashes = this.hashFunctions.map(_(str))
+    val newArr = hashes.foldLeft(this.array)((acc, hash) => acc.updated(hash, true))
+    BloomFilter(nbOfItems, falsePositiveProbability, newArr, hashFunctions)
+  }
+
+  def ++=(elems: Seq[A]): BloomFilter[A] = elems.foldLeft(this)(_ += _)
+
+  def contains(elem: A): Boolean = {
+    val str = elem.toString
+    this.hashFunctions.forall(fn => this.array(fn(str)))
+  }
+}
+
+object BloomFilter {
+
+  def apply[A](nbOfItems: Int, falsePositiveProbability: Double): BloomFilter[A] = {
+    val arraySize = Math.ceil(Math.abs(nbOfItems * Math.log(falsePositiveProbability)) / Math.log(1 / Math.pow(Math.log(2), 2))).toInt
+    val array = Array.ofDim[Boolean](arraySize)
+
+    val nbOfHashFunctions = Math.round((arraySize / nbOfItems) * Math.log(2)).toInt
+    val hashFunctions: List[String => Int] =
+      (1 to nbOfHashFunctions).toList.map(_ => {
+        val seed = Random.nextInt
```

Comment:
Might want to seed the `Random` with something like current time, to ensure the bloom filter is neutral. https://www.scala-lang.org/api/2.12.0/scala/util/Random.html#%3Cinit%3E(seed:Long):scala.util.Random

Reply:
Mh I don't understand, what do you mean by neutral?

Reply:
I mean, you're seeding the Murmur3 hash using this seed. But the seed itself is deterministic since `Random` is using some default seed. Unless you create one using current time (eg. `new Random(timeSinceEpoch)`)
On [wiki](https://en.m.wikipedia.org/wiki/Random_seed):
> A pseudorandom number generator's number sequence is completely determined by the seed: thus, if a pseudorandom number generator is reinitialized with the same seed, it will produce the same sequence of numbers.

Without a seed, `nextInt` will always produce the same sequence, which makes the Murmur3 hash always produce the same sequence. Not sure if neutral is the right word. But basically since there's a probability of false positive, we should avoid being predictable/biased.

Reply:
If nextInt, in my case, was always returning the same number, all the hash functions would be identical and my bloom filter wouldn't work :smile: I have empirically confirmed thaat the hash functions are all different.
My assumption is that, if you don't give it a seed, it goes a fetch one from the host's entropy.
Your quote talks about using the same seed, how about using no seed though?
> Without a seed, nextInt will always produce the same sequence

Where did you get this from?

Reply:
Ah, did not know that the `util.Random` library fetches a default seed when you import it. This is not true in C/C++ IIRC. No problem then 😅

Reply:
http://www.cplusplus.com/reference/cstdlib/rand/?kw=rand
where I got this assumption, if you're curious

Reply:
Well the way I see it, it doesn't even "fetch a seed when you import it", but goes fetch random numbers from the system directly when not given a seed. Again, that's my assumption.
In Haskell iirc it's quite similar: you have two ways to get random numbers. Either deterministically with a "root" random number and then seeds, or you also have another method that goes fetch a new random number from the host every time.